sllm/serve/backends/transformers_backend.py

```diff
@@ -180,56 +215,199 @@ async def generate(self, request_data: Optional[Dict[str, Any]]):
+        with self.status_lock:
+            if self.status == BackendStatus.DELETING:
+                return
+            self.status = BackendStatus.DELETING
+            if self.inf_status:
+                self.inf_status.status = BackendStatus.DELETING
 
-        # Simulate token counts for the response
-        prompt_tokens = len(self.tokenizer.tokenize(prompt))
-        completion_tokens = len(self.tokenizer.tokenize(output_text))
-        total_tokens = prompt_tokens + completion_tokens
+        while self.inf_status and len(self.inf_status.get()) > 0:
+            logger.info("Waiting for all requests to finish")
+            time.sleep(1)
 
-        # Generate response compatible with OpenAI's API
-        response = {
-            "id": f"chatcmpl-{uuid.uuid4()}",
-            "object": "chat.completion",
-            "created": int(time.time()),
-            "model": model_name,
-            "choices": [
-                {
-                    "index": 0,
-                    "message": {"role": "assistant", "content": output_text},
-                    "logprobs": None,
-                    "finish_reason": "stop",
-                }
-            ],
-            "usage": {
-                "prompt_tokens": prompt_tokens,
-                "completion_tokens": completion_tokens,
-                "total_tokens": total_tokens,
-            },
-        }
+        if self.model is not None:
+            del self.model
 
-        return response
+    def stop(self) -> None:
+        """Wait for all requests to finish and shutdown the backend."""
+        with self.status_lock:
+            if self.status.value >= BackendStatus.STOPPING.value:
+                return
+            self.status = BackendStatus.STOPPING
+        while self.inf_status and len(self.inf_status.get()) > 0:
+            logger.info("Waiting for all requests to finish")
+            time.sleep(1)
+        logger.info("All requests finished. Shutting down the backend.")
+        self.shutdown()
+
+    def get_current_tokens(self):
+        """Return a list of all ongoing request tokens."""
+        with self.status_lock:
+            if self.status != BackendStatus.RUNNING:
+                return []
+
+        status = self.inf_status.get()
+        logger.info(f"Current tokens: {status}")
+        return status
+
+    def resume_kv_cache(self, request_datas):
+        logger.info(f"Resuming cache for {request_datas}")
+        with torch.no_grad():
+            input_ids = torch.tensor(request_datas).reshape(1, -1).to("cuda")
+            output = self.model.generate(
+                input_ids,
+                past_key_values=self.past_key_values,
+                max_new_tokens=1,
+                return_dict_in_generate=True,
+                return_legacy_cache=True,
+            )
+            self.past_key_values = output.past_key_values
+        logger.info(f"Resumed {len(self.past_key_values[0][0][0][0])} tokens")
+
```

Comment:
This function should return the generated new token id, for later usage in `resume_generate`

Reply:
Hi! This function is designed to resume the KV cache for tokens already generated on the source node. It updates `self.past_key_values` with the resumed KV cache, which is then used as shown here:  
https://github.com/ServerlessLLM/ServerlessLLM/blob/114fb695ea907773b7c35af788c688e3e9447409/sllm/serve/backends/transformers_backend.py#L365-L371  

Since the function's purpose is to handle the KV cache update, it doesn’t seem necessary to return anything. What do you think?

Reply:
I have a question as well. The paper mentions that only tokens are transferred, and the KV cache is recomputed through prefilling. However, the code appears to indicate that the KV cache will be transferred instead.

Reply:
> I have a question as well. The paper mentions that only tokens are transferred, and the KV cache is recomputed through prefilling. However, the code appears to indicate that the KV cache will be transferred instead.

Could you share the specific code snippet that seems to indicate the KV cache is being transferred? It will help clarify any misunderstandings and provide a more accurate explanation. Thanks!

Reply:
I think it looks fine for me not to return anything here as long as the `resume_generate` bug can be handled on the caller side, i.e., write a check logic, asserting the `current_output` is at least one token longer than `self.past_key_values`